### DISCUSSION TITLE
Are there any decent logging libraries that work with the edge runtime?

### DISCUSSION THREAD

**Discussion by rinald-ta** [vercel/next.js#67213] [App Router] [answered]
I've had a hard time adding some form of logging to Next.js middleware when self-hosting. I've tried to use `winston` and `pino` since those are the most popular options, but both of them have issued when used in middleware, or any route that uses the edge runtime. (`winston` uses `process.nextTick` internally, which is not supported, and `pino` transports don't seem to work at all: `TypeError: pino__WEBPACK_IMPORTED_MODULE_0__.pino.transport is not a function`)

Are there any decent logging libraries that work with the edge runtime?

**Comment by leerob** [comment, marked as answer]
Good news, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental

**Reply by rinald-ta** [reply]
That sounds pretty exciting @leerob! I'll make sure to test it out.
Still, that doesn't quite answer my original question. I was looking for any recomendations that are production-ready. I would also consider the `after` API for my specific use case: https://nextjs.org/docs/app/api-reference/functions/after
